src/Workspaces/Core/Portable/Workspace/Workspace_Events.cs

```diff
@@ -59,22 +78,40 @@ protected Task RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind kind, Solutio
             projectId = documentId.ProjectId;
         }
 
-        var ev = GetEventHandlers<WorkspaceChangeEventArgs>(WorkspaceChangeEventName);
+        var ev = GetEventHandlers<WorkspaceChangeEventArgs>(WorkspaceChangedImmediateEventName);
+        WorkspaceChangeEventArgs args = null;
+        if (ev.HasHandlers)
+        {
+            args = new WorkspaceChangeEventArgs(kind, oldSolution, newSolution, projectId, documentId);
+            RaiseEventForHandlers(ev, args, isImmediate: true);
+        }
+
+        ev = GetEventHandlers<WorkspaceChangeEventArgs>(WorkspaceChangeEventName);
         if (ev.HasHandlers)
         {
             return this.ScheduleTask(() =>
             {
-                using (Logger.LogBlock(FunctionId.Workspace_Events, (s, p, d, k) => $"{s.Id} - {p} - {d} {kind.ToString()}", newSolution, projectId, documentId, kind, CancellationToken.None))
-                {
-                    var args = new WorkspaceChangeEventArgs(kind, oldSolution, newSolution, projectId, documentId);
-                    ev.RaiseEvent(static (handler, arg) => handler(arg.self, arg.args), (self: this, args));
-                }
+                args ??= new WorkspaceChangeEventArgs(kind, oldSolution, newSolution, projectId, documentId);
+                RaiseEventForHandlers(ev, args, isImmediate: false);
             }, WorkspaceChangeEventName);
         }
         else
         {
             return Task.CompletedTask;
         }
+
+        static void RaiseEventForHandlers(
+            EventMap.EventHandlerSet<EventHandler<WorkspaceChangeEventArgs>> handlers,
+            WorkspaceChangeEventArgs args,
+            bool isImmediate)
+        {
+            var functionId = isImmediate ? FunctionId.Workspace_EventsImmediate : FunctionId.Workspace_Events;
```

Comment:
Consider just passing the funding id along instead of the bool

Reply:
done!